hathor/consensus/poa/poa_block_producer.py

```diff
@@ -0,0 +1,172 @@
+#  Copyright 2024 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from structlog import get_logger
+from twisted.internet.task import LoopingCall
+
+from hathor.conf.settings import HathorSettings
+from hathor.consensus import poa
+from hathor.consensus.consensus_settings import PoaSettings
+from hathor.crypto.util import get_public_key_bytes_compressed
+from hathor.reactor import ReactorProtocol
+from hathor.transaction.poa import PoaBlock
+
+if TYPE_CHECKING:
+    from hathor.consensus.poa import PoaSigner
+    from hathor.manager import HathorManager
+
+logger = get_logger()
+
+# Number of seconds to wait for a sync to finish before trying to produce blocks
+_WAIT_SYNC_DELAY: int = 30
+
+# Number of seconds used in random delay calculation
+_RANDOM_DELAY_MULTIPLIER: int = 1
+
+
+class PoaBlockProducer:
+    """
+    This class is analogous to mining classes, but for Proof-of-Authority networks.
+    It waits for blocks to arrive, gets templates, and propagates new blocks accordingly.
+    """
+    __slots__ = (
+        '_log',
+        '_settings',
+        '_poa_settings',
+        '_reactor',
+        '_manager',
+        '_poa_signer',
+        '_signer_index',
+        '_started_producing',
+        '_start_producing_lc',
+        '_schedule_block_lc',
+        '_last_seen_best_block',
+    )
+
+    def __init__(
+        self,
+        *,
+        settings: HathorSettings,
+        reactor: ReactorProtocol,
+        manager: 'HathorManager',
+        poa_signer: PoaSigner,
+    ) -> None:
+        assert isinstance(settings.CONSENSUS_ALGORITHM, PoaSettings)
+        self._log = logger.new()
+        self._settings = settings
+        self._poa_settings = settings.CONSENSUS_ALGORITHM
+        self._reactor = reactor
+        self._manager = manager
+        self._poa_signer = poa_signer
+        self._signer_index = self._calculate_signer_index(self._poa_settings, self._poa_signer)
+        self._last_seen_best_block: PoaBlock | None = None
+
+        self._started_producing = False
+        self._start_producing_lc = LoopingCall(self._start_producing)
```

Comment:
@msbrogli commented on https://github.com/HathorNetwork/hathor-core-side-dag/pull/11#discussion_r1634983595:

> As it is a single execution to flip the `self._started_producing` flag, it seems safe to use `callLater()`.

Reply:
@msbrogli I would still need to recurse using `callLater()` until the node is synced, and using the same argument, the loop would be stopped if for any reason `manager.can_start_mining()` raises an exception.

Reply:
You're right. Ok. Thanks!